### ISSUE TITLE
GDScript 2.0: `@onready @export var n: Node = $Default` breaks.

### ISSUE THREAD

**Issue by mieldepoche** [godotengine/godot#62741]
### Godot version

4.0.alpha11

### System information

Linux

### Issue description

Consider the following line:
```gdscript
@onready @export var n: Node = $Node2D:
```
This is an export for a `Node` with a default value. The default value forces me to use `onready`.
However due to initialisation order, if I set a custom export value in the inspector, the following happens:
 1. the exported value is set on init
 2. since the variable is `onready`, the *default* value is then set on ready, overriding the exported one, meaning you can't actually export.

Even worse, if a setter is defined the setter is called when setting the exported value before simply overriding this value with the default one and without calling the setter:
```gdscript
extends Node3D

@onready @export var n: Node = $Foo:
	set(value):
		print("setter n was ", n, " now ", value)
		n = value

func _init() -> void:
	print("init ", n)

func _ready() -> void:
	print("ready ", n)

```
![2022-07-05_18-13-26 export onready hell](https://user-images.githubusercontent.com/71937306/177371707-4fcca2a9-c3f7-496b-9754-c7043afe89af.png)

prints:
```
init null
setter n was null now Bar:[Node2D:24763171212]
ready Foo:[Node2D:24746393997]
```

### Steps to reproduce

```gdscript
extends Node

@onready @export var five := 5

func _ready() -> void:
	print(five)

```

**Comment by adamscott** [comment]
Hi @mieldepoche! Sorry for the delay, I was combing through the GDScript issues today when I stumbled upon your issue.

Just a quick note: I had trouble understanding the issue, but I think I was able to create a minimal reproduction project that represents what you were trying to explain:
[62741.zip](https://github.com/godotengine/godot/files/10368831/62741.zip)

But please, especially when the code seems simple to reproduce but involves creating and assigning nodes, please create and join a MRP to the issue. It saves time for the contributors and you make sure that they are on the same page of you.

**Comment by adamscott** [comment]
Personally, I think that it doesn't break, that it does exactly what it says on the tin.

The `@onready` annotation is defined as this [in the docs](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_basics.html#onready-annotation).

>```gdscript
>var my_label
>
>func _ready():
>    my_label = get_node("MyLabel")
>```
>
>This can get a little cumbersome, especially when nodes and external references pile up. For this, GDScript has the @&NoBreak;onready annotation, that defers initialization of a member variable until _ready() is called. It can replace the above code with a single line:
>
>```gdscript
>@onready var my_label = get_node("MyLabel")
>```

Is this frustrating to use? Maybe. I suggest you to rather add inside your `_ready()` method:
```gdscript
extends Node3D

@export var n: Node

func _ready():
    if n == null:
        n = $Foo
```

**Comment by adamscott** [comment]
@anvilfolk just made me realize that every instance of `@onready @export` would overwrite the exported value with the defined value, no matter the circumstances.

Maybe `@onready` should always check if a value was assigned first before assigning the value.

**Comment by anvilfolk** [comment]
I only think this applies to nodes. Can anyone think of another situation where `@onready @export@` would be used together? The documentation describes the usage of `@onready` as basically to obtain node references at runtime.

For example, I don't think the following scenario makes any sense.
```GDScript
@onready @export some_var_from_child : int = $child_node.some_var
```

There is no point in the `@export` because either 1) `$child_node` exists, and therefore the default value should be overwritten, or 2) `$child_node` does not exist and the code explodes.

So basically, I think this issue describes the only situation in which both `@onready @export` annotations would be used. So if there's only one use case, nothing else is broken by changing current behavior, _and_ current behavior is unintuitive and unhelpful, it probably makes sense to have it behave as expected?

**Comment by anvilfolk** [comment]
I guess the alternative is to do what you did in 3.5, which is `@export` a `NodePath` and `@onready` a `Node` based on the exported NodePath - but the whole point of being able to use nodes is to avoid that mess :)

**Comment by akien-mga** [comment]
This is partially fixed by #72487.

> Fixes (kind of) https://github.com/godotengine/godot/issues/62741 by making this case an error by default. I believe we can discuss better behavior for a later release (which might break compatibility but get to be actually useful).

Moving to 4.x milestone then as the immediate problem is addressed for 4.0.

**Comment by YuriSizov** [comment]
Previous PR was reverted, and then superseded by https://github.com/godotengine/godot/pull/72608. The same note as above applies.

**Comment by HolonProduction** [comment]
I'm going to close this. It behaves consistently with what those annotations do and we have a warning about combining them which is set to error by default.
